packages/website/src/components/hooks/useHashState.ts

```diff
@@ -0,0 +1,163 @@
+        searchParams.get('sourceType') === 'script'
+          ? 'script'
+          : 'module',
+      code: searchParams.has('code')
+        ? readQueryParam(searchParams.get('code')!)
+        : '',
+      rules: searchParams.has('rules')
+        ? (JSON.parse(
+            readQueryParam(searchParams.get('rules')!),
+          ) as RulesRecord)
+        : undefined,
+      tsConfig: searchParams.has('tsConfig')
+        ? (JSON.parse(
+            readQueryParam(searchParams.get('tsConfig')!),
+          ) as CompilerFlags)
+        : undefined,
+    };
+  } catch (e) {
+    console.warn(e);
+  }
+  return undefined;
+};
+
+const writeStateToUrl = debounce(
+  (newState: ConfigModel, refresh = false): void => {
+    if (typeof window === 'undefined') {
+      return;
+    }
+    try {
+      const json: string = Object.entries({
+        ts: newState.ts,
+        jsx: newState.jsx,
+        sourceType: newState.sourceType,
+        showAST: newState.showAST,
+        code: newState.code ? writeQueryParam(newState.code) : undefined,
+        rules: newState.rules
+          ? writeQueryParam(JSON.stringify(newState.rules))
+          : undefined,
+        tsConfig: newState.tsConfig
+          ? writeQueryParam(JSON.stringify(newState.tsConfig))
+          : undefined,
+      })
+        .filter(item => item[1])
+        .map(item => `${encodeURIComponent(item[0])}=${item[1]}`)
+        .join('&');
+
+      if (refresh) {
+        window.location.replace(`${window.location.pathname}#${json}`);
+        window.location.reload();
+      } else {
+        window.history.pushState(
+          undefined,
+          document.title,
+          `${window.location.pathname}#${json}`,
+        );
+      }
+    } catch (e) {
+      console.warn(e);
+    }
+  },
+  100,
+);
+
+type HashAction<T extends ConfigModel> =
+  | { key: keyof T; value: T[keyof T] }
+  | T;
+
+function hashReducer(
+  prevState: ConfigModel,
+  action: HashAction<ConfigModel>,
+): ConfigModel {
+  if ('key' in action && 'value' in action) {
+    const newState = { ...prevState, [action.key]: action.value };
+    writeStateToUrl(newState, action.key === 'ts');
+    return newState;
+  } else {
+    return action;
+  }
+}
+
+type Args<T = ConfigModel, X extends keyof T = keyof T> =
+  | [T?]
+  | [key: X, value: T[X]];
+
+interface HashStateFn {
+  <X extends ConfigModel>(value?: X): void;
+  <X extends keyof ConfigModel>(key: X, value: ConfigModel[X]): void;
+}
+
+function useHashState(initialState: ConfigModel): [ConfigModel, HashStateFn] {
+  const [state, setState] = useReducer(hashReducer, initialState);
+
+  const onHashChange = (): void => {
+    const parsedState = parseStateFromUrl();
+    if (parsedState) {
+      setState(parsedState);
+    }
+  };
+
+  useEffect(() => {
+    window.addEventListener('hashchange', onHashChange);
+    return (): void => {
+      window.removeEventListener('hashchange', onHashChange);
+    };
+  }, []);
```

Comment:
Nit: you might consider `react-use`'s `useEvent` here.

Reply:
i don't like libraries that are doing to simple thinks, its overkill and adding it just to replace seem like overkill
```ts
  useEffect(() => {
    window.addEventListener('resize', resize);
    return (): void => {
      window.removeEventListener('resize', resize);
    };
  });
```